src/main/java/microsoft/exchange/webservices/data/core/EwsUtilities.java

```diff
@@ -875,79 +870,50 @@ public static TimeSpan getXSDurationToTimeSpan(String xsDuration) {
-    int day = 0;
-    if (m.find()) {
-      day = Integer.parseInt(m.group().substring(0,
-          m.group().indexOf("D")));
-    }
-
-    // Hour
-    m = PATTERN_HOUR.matcher(xsDuration);
-    int hour = 0;
-    if (m.find()) {
-      hour = Integer.parseInt(m.group().substring(0,
-          m.group().indexOf("H")));
-    }
-
-    // Minute
-    m = PATTERN_MINUTES.matcher(xsDuration);
-    int minute = 0;
-    if (m.find()) {
-      minute = Integer.parseInt(m.group().substring(0,
-          m.group().indexOf("M")));
+    // Removing leading '-'
+    if (negative) {
+      xsDuration = xsDuration.replace("-P", "P");
     }
 
-    // Seconds
-    m = PATTERN_SECONDS.matcher(xsDuration);
-    int seconds = 0;
-    if (m.find()) {
-      seconds = Integer.parseInt(m.group().substring(0,
-          m.group().indexOf(".")));
-    }
+    Period period = Period.parse(xsDuration, ISOPeriodFormat.standard());
+    int year = period.getYears();
+    int month = period.getMonths();
+    int day = period.getDays();
+    int hour = period.getHours();
+    int minute = period.getMinutes();
+    int seconds = period.getSeconds();
+    int milliseconds = period.getMillis();
+
+    StringBuilder debugMessage = new StringBuilder();
+    debugMessage.append("negative: " + negative);
+    debugMessage.append(" / ");
+    debugMessage.append("year: " + year);
+    debugMessage.append(" / ");
+    debugMessage.append("month: " + month);
+    debugMessage.append(" / ");
+    debugMessage.append("day: " + day);
+    debugMessage.append(" / ");
+    debugMessage.append("hour: " + hour);
+    debugMessage.append(" / ");
+    debugMessage.append("minute: " + minute);
+    debugMessage.append(" / ");
+    debugMessage.append("seconds: " + seconds);
+    debugMessage.append(" / ");
+    debugMessage.append("milliseconds: " + milliseconds);
+
+    LOG.debug("Parded TimeSpan: " + debugMessage.toString());
 
-    int milliseconds = 0;
-    m = PATTERN_MILLISECONDS.matcher(xsDuration);
-    if (m.find()) {
-      // Only allowed 4 digits of precision
-      if (m.group().length() > 5) {
-        milliseconds = Integer.parseInt(m.group().substring(0, 4));
-      } else {
-        seconds = Integer.parseInt(m.group().substring(0,
-            m.group().indexOf("S")));
-      }
-    }
-
-    // Apply conversions of year and months to days.
-    // Year = 365 days
-    // Month = 30 days
     day = day + (year * 365) + (month * 30);
-    // TimeSpan retval = new TimeSpan(day, hour, minute, seconds,
-    // milliseconds);
-    long retval = (((((((day * 24) + hour) * 60) + minute) * 60) +
-        seconds) * 1000) + milliseconds;
+    
+    long retval =
+        day * TimeSpan.DAYS + hour * TimeSpan.HOURS + minute * TimeSpan.MINUTES + seconds
+            * TimeSpan.SECONDS + milliseconds * TimeSpan.MILLISECONDS;
+    
     if (negative) {
       retval = -retval;
     }
-    return new TimeSpan(retval);
 
+    return new TimeSpan(retval);
```

Comment:
@serious6 I had a closer at Joda Time Period. It seems, that it does not support duration with years and months, which makes sense, because year and months have different length. Here we have two methods: getXSDurationToTimeSpan(), which tries to include years and months, and getXSDurationToTimeSpanValue(), which is not. 
Is there any particular reason for having two methods or I can just merge leave one method getXSDurationToTimeSpanValue() and refactor usage of getXSDurationToTimeSpan() in other classes?